### PR TITLE
Arm64 irq1

### PR DESCRIPTION
**Pull request by lipengfei28**
## Summary
arm64: simply the vectors

### CODE REVIEW

**Comment by lipengfei28** [comment]
1. save FPU regs enter exception and retore exit exception every time
    2. simp the vectors
    3. save sp_el0 enter exception and restore sp_el0 exit exception every time

**Review by acassis — APPROVED** [review]

File: arch/arm64/src/common/arm64_syscall.c [hunk 1]
```diff
@@ -192,13 +192,13 @@ uint64_t *arm64_syscall_switch(uint64_t * regs)
...
 
-          ret_regs = (uint64_t *)f_regs->regs[REG_X1];
+          ret_regs = (uint64_t *)regs[REG_X1];
 
           DEBUGASSERT(ret_regs);
         }
         break;
 
-      /* x0 = SYS_switch_context:  This a switch context command:
+       /* x0 = SYS_switch_context:  This a switch context command:
```
[the remark this answers is not in the document]
Comment (lipengfei28) [reply arch/arm64/src/common/arm64_syscall.c]: done

File: arch/arm64/src/common/arm64_syscall.c [hunk 2]
```diff
@@ -216,85 +216,14 @@ uint64_t *arm64_syscall_switch(uint64_t * regs)
 
       case SYS_switch_context:
         {
-          DEBUGASSERT(f_regs->regs[REG_X1] != 0 &&
-                      f_regs->regs[REG_X2] != 0);
-          *(uint64_t **)f_regs->regs[REG_X1] = regs;
+          DEBUGASSERT(regs[REG_X1] != 0 &&
+                      regs[REG_X2] != 0);
```
[the remark this answers is not in the document]
Comment (lipengfei28) [reply arch/arm64/src/common/arm64_syscall.c]: done

File: arch/arm64/src/common/arm64_arch.h [hunk 3]
```diff
@@ -394,6 +394,26 @@ static inline void arch_nop(void)
   __asm__ volatile ("nop");
 }
 
+/****************************************************************************
+ * Name:
+ *   arm64_current_el()
+ *
+ * Description:
+ *
+ *   Get current execute level
```
Comment (acassis) [inline arch/arm64/src/common/arm64_arch.h:403]:
```suggestion
 *   Get current execution level

**Review by xiaoxiang781216 — APPROVED** [review]

File: arch/arm64/src/common/arm64_vectors.S [hunk 4]
```diff
@@ -196,179 +139,32 @@ SECTION_FUNC(text, arm64_jump_to_user)
...
-#endif
-    /* 0x15 = SVC system call */
-
-    cmp    x10, #0x15
-
-    /* if this is a svc call ?*/
-
-    bne    exc_handle
-
-#ifdef CONFIG_LIB_SYSCALL
```
Comment (pussuw) [inline arch/arm64/src/common/arm64_vectors.S:229]: Why exactly was this part changed??? Now you destroy the incoming system call parameter registers, and then re-read them from the context wasting massive amounts of CPU time??

WHY WAS THIS CHANGED?

File: arch/arm64/src/common/arm64_syscall.c [hunk 5]
```diff
@@ -397,11 +324,59 @@ int arm64_syscall(uint64_t *regs)
...
-          break;
+          /* Verify that the SYS call number is within range */
+
+          DEBUGASSERT(cmd >= CONFIG_SYS_RESERVED && cmd < SYS_maxsyscall);
+
+          /* Make sure that there is a no saved SYSCALL return address.  We
+           * cannot yet handle nested system calls.
+           */
+
+          regs[REG_X0] = dispatch_syscall(regs[REG_X0], regs[REG_X1],
```
Comment (pussuw) [inline arch/arm64/src/common/arm64_syscall.c:344]: This is completely unnecessary, the registers x0-x7 are already correct when coming from userspace. There is NO REASON TO READ THEM FROM THE CONTEXT AREA. What I had done is much more optimized than what this does.

Comment (pussuw) [reply arch/arm64/src/common/arm64_syscall.c:344]: Another thing this change does; the system call now runs with interrupts disabled, affecting interrupt service latency.

**Comment by jlaitine** [comment]
After this PR got merged, nuttx is not compiling any more. I am getting this:

common/arm64_fatal.c:84:30: error: initialized field overwritten [-Werror=override-init]
   84 |   [ESR_ELX_EC_UNKNOWN]     = "Unknown/Uncategorized",
      |                              ^~~~~~~~~~~~~~~~~~~~~~~
common/arm64_fatal.c:84:30: note: (near initialization for 'g_esr_class_str[0]')
common/arm64_fatal.c:85:30: error: initialized field overwritten [-Werror=override-init]
   85 |   [ESR_ELX_EC_WFX]         = "WFI/WFE",
      |                              ^~~~~~~~~
common/arm64_fatal.c:85:30: note: (near initialization for 'g_esr_class_str[1]')
common/arm64_fatal.c:86:30: error: initialized field overwritten [-Werror=override-init]
   86 |   [ESR_ELX_EC_CP15_32]     = "CP15 MCR/MRC",
      |                              ^~~~~~~~~~~~~~
common/arm64_fatal.c:86:30: note: (near initialization for 'g_esr_class_str[3]')
common/arm64_fatal.c:87:30: error: initialized field overwritten [-Werror=override-init]
   87 |   [ESR_ELX_EC_CP15_64]     = "CP15 MCRR/MRRC",
      |                              ^~~~~~~~~~~~~~~~
common/arm64_fatal.c:87:30: note: (near initialization for 'g_esr_class_str[4]')
common/arm64_fatal.c:88:30: error: initialized field overwritten [-Werror=override-init]
   88 |   [ESR_ELX_EC_CP14_MR]     = "CP14 MCR/MRC",
      |                              ^~~~~~~~~~~~~~
common/arm64_fatal.c:88:30: note: (near initialization for 'g_esr_class_str[5]')
common/arm64_fatal.c:89:30: error: initialized field overwritten [-Werror=override-init]
   89 |   [ESR_ELX_EC_CP14_LS]     = "CP14 LDC/STC",
      |                              ^~~~~~~~~~~~~~
common/arm64_fatal.c:89:30: note: (near initialization for 'g_esr_class_str[6]')
common/arm64_fatal.c:90:30: error: initialized field overwritten [-Werror=override-init]
   90 |   [ESR_ELX_EC_FP_ASIMD]    = "ASIMD",

File: arch/arm64/src/common/arm64_syscall.c (hunk 5 above)
Comment (xiaoxiang781216) [reply arch/arm64/src/common/arm64_syscall.c:344]: @lipengfei28 please take a look.

**Comment by lipengfei28** [comment]
commit 8288fe4e305bb905fa3ef18d24f895cf5ca33359 (HEAD -> master, apache/master)
Author: chao an <anchao@lixiang.com>
Date:   Mon Sep 23 08:59:15 2024 +0800

    Revert "toolchain/ghs: Fix CONFIG_SCHED_CRITMONITOR_MAXTIME_XXX "zero used for undefined preprocessing identifier" warnings"
    
    move private define to public
    
    This reverts commit 236678d730220138416724d2cab25d6e1bd2ce9d.
    
    Signed-off-by: chao an <anchao@lixiang.com>

commit a5b85fcd961fb8a62fb31eb88a74f4c64406a2d7
Author: wangjianyu3 <wangjianyu3@xiaomi.com>
Date:   Mon Sep 23 18:44:17 2024 +0800

    misc/rpmsgdev: Fix invalid pointer error when there are more than one remotes
    
    Test: (see tests/testcases/rpmsgdev for details)
      # 1. Register dummy device
      testdev -d 0 -r "/dev/ttyGNSS0"
      # 2. Call rpmsgdev_export() to export the device to remote
      testdev -d 2 -c "droid" -l "/dev/ttyGNSS0"
    
    Log:
      [ap] arm_busfault: PANIC!!! Bus Fault:
      [ap] arm_busfault:        IRQ: 5 regs: 0x3c434e44
      [ap] arm_busfault:        BASEPRI: 00000000 PRIMASK: 00000000 IPSR: 00000005 CONTROL: 00000004
      [ap] arm_busfault:        CFSR: 00008200 HFSR: 00000000 DFSR: 00000000 BFAR: 00000000 AFSR: 00040000
      [ap] arm_busfault: Bus Fault Reason:
      [ap] arm_busfault:        Precise data bus error
      [ap] dump_assert_info: Current Version: NuttX ****** ***** *** 12.3.0 ********** Sep 23 2024 18:35:50 arm
      [ap] dump_assert_info: Assertion failed panic: at file: armv8-m/arm_busfault.c:113 task: testdev process: testdev 0x2c86ca75
    
This commit build is ok

**Comment by jlaitine** [comment]
This PR seems problematic also for the performance; what is the reason for example for saving the FPU registers every time? Before this PR we had well working and performant arm64 in both flat and kernel builds (and extensively using FPU as well).

Or why to execute system calls with interrupts disabled? That is clearly wrong.

Can you please explain what is the problem this PR is trying to solve? Just to break everything for fun?

Sorry to be negative, but I don't see any benefits in breaking things like this, please explain!

**Comment by jlaitine** [comment]
FYI: I managed to get things running in our own branches again by reverting everything in this PR except for e38f2b2a6d

File: arch/arm64/src/common/arm64_vector_table.S [hunk 6]
```diff
@@ -78,6 +78,10 @@
 #endif
     stp    x0, x1,  [sp, #8 * REG_ELR]
 
+    mrs    x0, sp_el0
+    mrs    x1, tpidrro_el0
+    stp    x0, x1,  [sp, #8 * REG_SP_EL0]
+
```
Comment (jlaitine) [inline arch/arm64/src/common/arm64_vector_table.S:84]: This patch is completely useless. Kernel doesn't use sp_el0, you are only suppposed to store it when changing context. Like it was done before this PR.

**Comment by xiaoxiang781216** [comment]
let's revert this patch first and refine the change later.

**Comment by jlaitine** [comment]
The idea in "arm64: refine the fatal handler" is good; I'd suggest making a separate PR of that one, on top of the existing implementation.

**Comment by xiaoxiang781216** [comment]
@GUIDINGLI will resend the patch again.

**Comment by GUIDINGLI** [comment]
@jlaitine @pussuw 
https://github.com/apache/nuttx/pull/13652

**Comment by pussuw** [comment]
> @jlaitine @pussuw #13652

I'll take a look tomorrow, thanks

File: arch/arm64/src/common/arm64_vectors.S [hunk 7]
```diff
@@ -139,164 +139,33 @@ SECTION_FUNC(text, arm64_jump_to_user)
 
 GTEXT(arm64_sync_exc)
 SECTION_FUNC(text, arm64_sync_exc)
-    /* checking the EC value to see which exception need to be handle */
-
-#if CONFIG_ARCH_ARM64_EXCEPTION_LEVEL == 3
-    mrs    x0, esr_el3
+    /* Switch to IRQ stack and save current sp on it. */
+#ifdef CONFIG_SMP
```
Comment (pussuw) [inline arch/arm64/src/common/arm64_vectors.S:143]: @GUIDINGLI I assume this is the code that SMP needs ? I can adapt this part so that kernel mode works like before. I'll try to provide a patch tomorrow, or on Monday.

File: arch/arm64/src/common/arm64_arch.h [hunk 8]
```diff
@@ -299,6 +299,9 @@ struct regs_context
   uint64_t  spsr;
   uint64_t  sp_el0;
   uint64_t  exe_depth;
+#ifdef CONFIG_ARCH_FPU
```
Comment (pussuw) [inline arch/arm64/src/common/arm64_arch.h:302]: Is the content of this patch needed by SMP too ?

File: arch/arm64/src/common/arm64_vectors.S (hunk 7 above)
Comment (GUIDINGLI) [reply arch/arm64/src/common/arm64_vectors.S:143]: good

File: arch/arm64/src/common/arm64_arch.h (hunk 8 above)
Comment (GUIDINGLI) [reply arch/arm64/src/common/arm64_arch.h:302]: The removable of FPU trap reason:
1. The FPU trap logic is not stable in some SMP situation (HERE haven't the detail analysis)
2. Make the logic Complex and different to read.
3. The most important reason: it has almost no benefit for the performance, because the compiler will optimize the code with float/double instructions when use the option -O3 (always be). Even the printf/syslog will use use the FPU, so the FPU is used too often and almost each context switch.
Or, you must use -mgeneral-regs-only to limit the FPU registers usage, that will harmful for the speed.
5. How about use the FPU registers in the IRQ/SVC ISR, maybe there are hidden bugs ?
